### PR TITLE
Adds scores, scoreboard and combo counter to ddr mode

### PR DESCRIPTION
**Pull request by Ruzihm**

### CODE REVIEW

File: firmware/user/modes/mode_ddr.c [hunk 1]
```diff
@@ -231,39 +263,52 @@ void ICACHE_FLASH_ATTR ddrEnterMode(void)
...
 }
 
-static void ICACHE_FLASH_ATTR ddrStartGame(int tempo, float eighthNoteProbabilityModifier, int restAvoidanceProbability)
+static void ICACHE_FLASH_ATTR ddrGoToScores()
+{
+    ddr->mode = DDR_HIGHSCORE;
+    ddr->isNewHighScore = false;
+    ddr->currentDifficultyType = DDR_HARD;
+    timerDisarm(&ddr->TimerAdvanceHighScores);
+    timerArm(&ddr->TimerAdvanceHighScores, HIGH_SCORE_PERIOD, true);
```
Comment (AEFeinstein) [inline firmware/user/modes/mode_ddr.c]: Oh, I get it now. I flashed your changes before reviewing the code and was mashing the buttons on the high score screen to get it to show easy and medium. Having buttons cycle the pages here is more intuitive than doing it on a timer. You can put "<" and ">" in the bottom left and right corners to indicate cycle-ability, like on the root menu screen

File: firmware/user/modes/mode_ddr.c [hunk 2]
```diff
@@ -580,7 +638,40 @@ static void ICACHE_FLASH_ATTR ddrHandleArrows(void)
 static void ICACHE_FLASH_ATTR ddrAnimateNotes(void* arg __attribute__((unused)))
 {
     ddr->NoteIdx = (ddr->NoteIdx + 1) % 4;
-    // testUpdateDisplay();
+}
+
+static void ICACHE_FLASH_ATTR ddrMakeScoreString(ddrWinResult_t curWin, char* scoreText, unsigned int size)
```
Comment (AEFeinstein) [inline firmware/user/modes/mode_ddr.c:635]: Have you thought about checking for a nonzero score when building the string? Having a bunch of zero scores with 'C' ratings the first time you check the scores is a little odd. I'd either not show a letter grade, or return an empty string.

File: firmware/user/modes/mode_ddr.c [hunk 3]
```diff
@@ -580,7 +638,40 @@ static void ICACHE_FLASH_ATTR ddrHandleArrows(void)
...
+            typeText = "AA";
+            break;
+        case DDR_ALL_PERFECT:
+            typeText = "AAA";
+            break;                
+    }
+
+    if (ddr->isNewHighScore)
+    {
+        ets_snprintf(scoreText, size, "*%-3s %09d*", typeText, curWin.score);
```
Comment (AEFeinstein) [inline firmware/user/modes/mode_ddr.c]: The Tom Thumb font is already not the most readable, and the leading zeros don't help in that regard. `*%-3s %-9d*` may be a little easier on the eyes.

File: firmware/user/modes/mode_ddr.c [hunk 4]
```diff
@@ -593,16 +684,58 @@ static void ICACHE_FLASH_ATTR ddrUpdateDisplay(void* arg __attribute__((unused))
...
+
+            ddrWinResult_t* currentDiffScores;
+            const char* diffName;
+
+            switch(ddr->currentDifficultyType)
+            {
+                default:
+                case DDR_HARD:
+                    currentDiffScores = ddr->highScores.hardWins;
+                    diffName = " HARD";
```
Comment (AEFeinstein) [inline firmware/user/modes/mode_ddr.c]: Any reason for the leading space before " HARD" and " EASY"? If it was for centering purposes, you can use `textWidth()` to measure it before plotting, then plot it at `(OLED_WIDTH - width) / 2`.

File: firmware/user/modes/mode_ddr.c [hunk 5]
```diff
@@ -705,6 +839,15 @@ static void ICACHE_FLASH_ATTR ddrUpdateDisplay(void* arg __attribute__((unused))
                          OLED_HEIGHT - 1, WHITE);
             }
 
+            if (ddr->currentCombo > 3)
+            {
+                char comboText[24];
+                ets_snprintf(comboText, sizeof(comboText),   "Combo!!  %04d", ddr->currentCombo);
+                plotText(15, 30, comboText, TOM_THUMB, WHITE); 
+                // TODO: use INVERSE instead ?
```
Comment (AEFeinstein) [inline firmware/user/modes/mode_ddr.c]: Should be fixed in my master if you want to

File: firmware/user/modes/mode_ddr.c [hunk 6]
```diff
@@ -804,10 +955,99 @@ static void ICACHE_FLASH_ATTR ddrCheckSongEnd()
...
+        case DDR_MEDIUM:
+            currentDiffScores = ddr->highScores.mediumWins;
+            break;
+        case DDR_EASY:
+            currentDiffScores = ddr->highScores.easyWins;
+            break;
+    }
+
+    ddr->isNewHighScore = false;
+    int replaceIndex;
```
Comment (AEFeinstein) [inline firmware/user/modes/mode_ddr.c]: I may be reading this incorrectly, but wouldn't you want to do an insertion, not a replacement, in order to keep all the high scores sorted?

**Comment by AEFeinstein** [comment]
I may be getting better at Stomp, but it just _felt_ better when testing!

File: firmware/user/modes/mode_ddr.c (hunk 3 above)
Comment (Ruzihm) [reply firmware/user/modes/mode_ddr.c]: This definitely is an improvement. I went ahead with `*%-3s %9d*` to keep the number right justified

File: firmware/user/modes/mode_ddr.c (hunk 6 above)
Comment (Ruzihm) [reply firmware/user/modes/mode_ddr.c]: It was a horribly implemented insertion. 😆 
I've rewritten it now

File: firmware/user/modes/mode_ddr.c (hunk 3 above)
Comment (AEFeinstein) [reply firmware/user/modes/mode_ddr.c]: Makes sense, and you even justified your decision

**Pull request merged**